### PR TITLE
block: precalculate hashes if enabled and use them during compaction (downloading)

### PR DESCRIPTION
**Pull request by GiedriusS**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Added a change to `objstore.DownloadDir` so that it would be possible to ignore certain directories. Added functions for calculating the hashes of files and directories. Used them inside of `shipper` & other places. Wired up the new `shipper` changes and all of the other writers to command-line parameters so that the users can optionally enable them. There should be no performance impact if none of the flags (the default) have been enabled.

## Verification

Unit tests. Manual testing. A new e2e test.

### CODE REVIEW

File: CHANGELOG.md
Comment (bwplotka) [inline CHANGELOG.md:35]: Does it infinitely save data on disk and never delete it? :thinking: This is what I read from this

File: cmd/thanos/compact.go
Comment (bwplotka) [inline cmd/thanos/compact.go:278]: I think it's time for 
```
defer func() {
   if err  != nil {
      cancel()
   }
}()
```
(:

File: CHANGELOG.md
Comment (bwplotka) [reply CHANGELOG.md:35]: >  So, if you, for example, use persistent volumes on k8s and your Thanos Compact crashes or fails to make an iteration properly then the last downloaded files are not wiped from the disk

From the code it looks like they ARE wiped in fact, no?

File: cmd/thanos/compact.go
Comment (bwplotka) [inline cmd/thanos/compact.go:468]: I am a fan of avoiding bool options. What about `hashFunction` setting when you set what kind of hash to use (: And by default it will be none?

File: cmd/thanos/compact.go
Comment (bwplotka) [inline cmd/thanos/compact.go:515]: `downloading some file twice`  is vague - does it mean Thanos downloads files couple of times without checking? Let's maybe not specify what's the purpose - there are many reasons. Mainly to ensure data integrity.

Can we at some point share what is that performance cost and during what moment? ( :

File: CHANGELOG.md
Comment (bwplotka) [reply CHANGELOG.md:35]: Ok I think we need to improve this (really 2x reduce) description, and just say "on-crash & recovery we don't delete data anymore but we are able to start up without redownload.

File: pkg/block/block.go
Comment (bwplotka) [inline pkg/block/block.go:50]: no need for full metadata really - just hashes, right?

File: pkg/block/block.go
Comment (bwplotka) [inline pkg/block/block.go:47]: This is bit confusing. Can we just have deleteDir function directly used and mock object storage?

File: pkg/block/metadata/hash.go
Comment (bwplotka) [inline pkg/block/metadata/hash.go:19]: It's not kind but maybe `function`?

```suggestion
type HashFunc string
```
or maybe type?

File: pkg/block/metadata/hash.go
Comment (bwplotka) [inline pkg/block/metadata/hash.go:28]:
```suggestion
	Func  HashFunc
```
? or type?

File: pkg/block/metadata/hash.go
Comment (bwplotka) [inline pkg/block/metadata/hash.go:58]: We have error - let's use this instead.

The reason is compatibility. How would you add new function? (: Rollout Thanos in exactly same moment in all clusters?

File: pkg/block/metadata/hash.go
Comment (bwplotka) [inline pkg/block/metadata/hash.go:100]: Why?

File: pkg/compact/compact.go
Comment (bwplotka) [inline pkg/compact/compact.go:901]: so we don't check halt. I think you mean any error, which is ok - as long it's not full disk error ;p

Comment (bwplotka) [reply pkg/compact/compact.go:901]: also everywhere else

File: pkg/objstore/objstore.go
Comment (pracucci) [inline pkg/objstore/objstore.go:215]: This is a generic implementation which just takes in input `ignoredPaths`. In my opinion, this function should know nothing about the fact it will be called with `ignoredPaths` based on content hash comparison. For this reason, I would generalise this message to not mention the hashing thing.

File: pkg/block/block.go
Comment (pracucci) [inline pkg/block/block.go:52]: No need for this `if`. The next `for` will be a no-op if `m.Thanos.Hashes` is empty.

File: pkg/block/metadata/hash.go
Comment (pracucci) [inline pkg/block/metadata/hash.go:64]: Given this is `CalculateHash` while iterating over a directory, I'm wondering if we should name this function `CalculateHashDir`

**Review by pracucci — COMMENTED** [review]
Thanks for working on this! The overall design LGTM. I agree with Bartek's comments plus I've added a couple of minor ones. Thanks!

File: pkg/block/metadata/hash.go
Comment (GiedriusS) [reply pkg/block/metadata/hash.go:100]: Well, the deletion mark could actually just be removed from here since it will never appear while a block is being produced. However, I think the `meta.json` file should be definitely excluded because there will be confusion because the hash will most likely change *after* adding it to `m.Thanos.Hashes`.

File: pkg/verifier/safe_delete.go
Comment (bwplotka) [inline pkg/verifier/safe_delete.go:66]: Why we need to do this?

File: pkg/shipper/shipper.go
Comment (bwplotka) [inline pkg/shipper/shipper.go:363]:
```suggestion
	meta.Thanos.Hashes, err = metadata.CalculateHashDir(updir, metadata.SHA256Func, metadata.NominalIgnoredFiles...)
```

File: pkg/block/metadata/hash.go
Comment (bwplotka) [reply pkg/block/metadata/hash.go:105]: I mean, I would just rename it tbh or just use `meta.json` explicitly somewhere. `Nominal` does not mean anything to me :hugs: 

But overall it makes sense to skip hashing meta.json, it would hard... to achieve (:

**Review by bwplotka — APPROVED** [review]
LGTM, just minor comments.

File: cmd/thanos/receive.go
Comment (prmsrswt) [inline cmd/thanos/receive.go:100]: As this flag is used in multiple places, maybe we should extract out it in a function in `cmd/thanos/flags.go`.

File: pkg/block/metadata/meta.go
Comment (prmsrswt) [inline pkg/block/metadata/meta.go:63]: I guess you mean, `// Hashes is a map of hashes of files inside the block.`

**Review by prmsrswt — APPROVED** [review]
Just some minor nits, everything else looks good to me!

File: cmd/thanos/receive.go
Comment (GiedriusS) [reply cmd/thanos/receive.go:100]: I agree, let's do it. However, how should we call it? Simply `--hash-func`? WDYT @prmsrswt, @bwplotka ?

File: pkg/block/metadata/meta.go
Comment (GiedriusS) [reply pkg/block/metadata/meta.go:63]: :smile: yes, indeed. Thank you for spotting this! :heart:

File: pkg/verifier/safe_delete.go
Comment (GiedriusS) [reply pkg/verifier/safe_delete.go]: Good point, I have removed this change.

File: cmd/thanos/receive.go
Comment (prmsrswt) [reply cmd/thanos/receive.go:100]: I guess `--hash-func` would work.

**Comment by prmsrswt** [comment]
One more thing that came to my mind was, after using this feature, if someone decides to switch the hashing algorithm, al previous hashes would become unusable and we will trigger a redownload even though we have the correct files.

I guess we can also store the hashing algorithm used to create the hash alongside it so that we can calculate and match hashes using that algorithm irrespective of what algorithm is currently selected. This will also help us in identifying the hashing algorithm at places like Block Viewer and then showing some info to the user.

**Comment by GiedriusS** [comment]
Working on rebasing this.

**Comment by GiedriusS** [comment]
@bwplotka @prmsrswt @pracucci I have rebased this on `master`. PTAL! :muscle: This version looks a bit cleaner to me because we now have that `Files` array in `meta.json`.

**Comment by GiedriusS** [comment]
I believe this is ready for another round of reviews. There are some potential cleanups to be had by exposing `gatherFileStats` and reusing it in `e2eutil` however that leads to circular dependencies which I'd like to solve in a separate PR as to not get distracted here. It's already a couple of months for this PR ;) 

I also played around with this locally. An excerpt from a `meta.json` file:

```json
                        {                                                                                                                                                                                  
                                "rel_path": "index",                                                                                                                                                       
                                "size_bytes": 212315,
                                "hash": {                                                            
                                        "Func": "SHA256",
                                        "Value": "c7fbe6c03c8c2a89e8417a6124fe37452b7b22a2b78934d9b25bd38815c81c1f"
                                }
                        },
                        {                                                                            
                                "rel_path": "meta.json"
                        }
```

I have double-checked that the hash is correct with `sha256sum`. PTAL @thanos-io/thanos-maintainers  :heart:

File: pkg/block/block.go
Comment (bwplotka) [inline pkg/block/block.go:67]:
```suggestion
			level.Error(logger).Log("msg", "failed to calculate hash when downloading; re-downloading", "relPath", fl.RelPath, "err", err)
```
I think we should have some lint to check wrong log usage (:

Comment (bwplotka) [reply pkg/block/block.go:67]: Added https://github.com/thanos-io/thanos/issues/3680

File: pkg/block/block.go
Comment (bwplotka) [inline pkg/block/block.go:62]:
```suggestion
		if fl.Hash == nil || fl.Hash == metadata.None || fl.RelPath == "" {
```

File: pkg/block/metadata/hash_test.go
Comment (bwplotka) [inline pkg/block/metadata/hash_test.go:14]: `smoke`



Did not hear that word for some time... ;p Smoke test is valid naming though (:

File: pkg/compact/compact.go
Comment (bwplotka) [inline pkg/compact/compact.go:889]: Same, let's inline

File: pkg/objstore/inmem.go
Comment (bwplotka) [inline pkg/objstore/inmem.go:28]: I would vote for just using MetricBucket wrapper instead. You can just use count metric (not needed per file) right?

The first reason is that we already have an implementation for tracking this. Secondly, it's because it's not testing bucket (by name), it's in mem bucket. While it's recommended for testing usage, let's keep it production-grade if we can (: 

WDYT?

File: pkg/objstore/objstore.go
Comment (bwplotka) [inline pkg/objstore/objstore.go:204]:
```suggestion
func DownloadDir(ctx context.Context, logger log.Logger, bkt BucketReader, originalSrc, src, dst string, ignorePaths ...string) error {
```

Comment (bwplotka) [reply pkg/objstore/objstore.go:204]: For compatibility and indication that those are optional 🤗

File: pkg/objstore/objstore.go
Comment (bwplotka) [inline pkg/objstore/objstore.go:216]: Do we need this log line? 🤔 Might be spammy? Maybe debug level?

File: pkg/runutil/runutil.go
Comment (bwplotka) [inline pkg/runutil/runutil.go:165]:
```suggestion
func DeleteAll(dir string, ignoreDirs ..string) error {
```

This would be clear enough IMO WDYT?

File: pkg/testutil/e2eutil/prometheus.go
Comment (bwplotka) [inline pkg/testutil/e2eutil/prometheus.go:495]:
```suggestion
		if err := filepath.Walk(blockDir, func(path string, info os.FileInfo, err error) error {
			if info.IsDir() {
				return filepath.SkipDir
			}
			paths = append(paths, path)
			return nil
		}); err != nil {
			return id, errors.Wrapf(err, "walking %s", dir)
		}
```

File: pkg/testutil/e2eutil/prometheus.go
Comment (bwplotka) [inline pkg/testutil/e2eutil/prometheus.go:347]: I would vote for consistency and getting hashFunc instead like everywhere.

**Review by bwplotka — CHANGES_REQUESTED** [review]
Not sure if this changed overtime a lot or it's just me being rested after some holidays, but I found more suggestions to make, sorry for that 🙈 

Thanks for this. It looks generally ok, just some style / readability suggestions, and one major question.

File: CHANGELOG.md
Comment (MalloZup) [inline CHANGELOG.md:18]: `nit` I wonder if `does a smart thing` belong to the changelog.

Imho I would just remove it in sake of formal language.

File: pkg/block/metadata/hash.go
Comment (MalloZup) [inline pkg/block/metadata/hash.go:45]: I guess here we could apply the Coding-style guide?
https://github.com/thanos-io/thanos/blob/master/docs/contributing/coding-style-guide.md#defers-dont-forget-to-check-returned-errors

File: pkg/compact/compact.go
Comment (MalloZup) [inline pkg/compact/compact.go:900]: this comment although appreciated, from my pov we could improve its readability.

Especially the 2nd line, has imho to much logic word. (or if it is not not, then) which maybe could simplified the sentence and improve readability imho

File: pkg/runutil/runutil.go
Comment (MalloZup) [reply pkg/runutil/runutil.go:165]: yep

**Review by MalloZup — COMMENTED** [review]
good work ! :+1:  I just left some minor comment and took the occasion to review from curiosity pov to learn more the code .. :grin:

File: cmd/thanos/tools_bucket.go
Comment (GiedriusS) [reply cmd/thanos/tools_bucket.go:486]: Because it doesn't make sense to add it for both writers _and_ readers. It would not mean anything for components that only read and would only raise confusion e.g. `bucket ls`. As for separating this into a helper or something - maybe. I had an idea that we could separate the uploading part from `Shipper` into a separate struct/package and use that inside of shipper + here. But that is a completely separate story that I think we can do later given that such a refactoring does not impact the end user at all.

File: CHANGELOG.md
Comment (GiedriusS) [reply CHANGELOG.md]: Removed!

File: cmd/thanos/compact.go
Comment (GiedriusS) [reply cmd/thanos/compact.go:295]: You are correct - we can take a step further and do not delete these at the beginning due to the reasons you've mentioned.

**Comment by GiedriusS** [comment]
@bwplotka @MalloZup I have fixed this PR according to your comments, added more extensive unit tests + added a e2e test which has caught some nasty bugs. I truly believe this is ready for prime time. PTAL! :hugs:

File: cmd/thanos/compact.go
Comment (bwplotka) [inline cmd/thanos/compact.go:395]: I don't think you need to do it, won't it work without casting in this case? 🤔

File: cmd/thanos/downsample.go
Comment (bwplotka) [inline cmd/thanos/downsample.go:171]: What if it's disk full issue?

File: pkg/block/block.go
Comment (bwplotka) [inline pkg/block/block.go:48]:
```suggestion
// we do not download it. We always re-download the meta file.
```

File: pkg/compact/compact.go
Comment (bwplotka) [inline pkg/compact/compact.go:887]: 👍🏽

File: pkg/compact/compact.go
Comment (bwplotka) [inline pkg/compact/compact.go:963]: I think we need to check err, no?

File: pkg/compact/compact.go
Comment (bwplotka) [inline pkg/compact/compact.go:964]: Let's add `msg` to tell what this is warning about (:

File: pkg/store/bucket_test.go
Comment (bwplotka) [inline pkg/store/bucket_test.go:1057]: There might be some time for Uploader struct , but not in this PR 🤗

**Review by bwplotka — APPROVED** [review]
Only minor logging nit, otherwise this looks good to me. Good tests and logic. 👍🏽  LGTM after this tiny log fix.

File: cmd/thanos/compact.go
Comment (GiedriusS) [reply cmd/thanos/compact.go:395]: Unfortunately but it wouldn't work without casting:
```
cannot use conf.hashFunc (variable of type string) as metadata.HashFunc value in argument to downsampleBucket
```

File: pkg/compact/compact.go
Comment (GiedriusS) [reply pkg/compact/compact.go:963]: Actually, I have added this while trying to debug e2e tests. It's not needed at all so I have removed it.

File: pkg/compact/compact.go
Comment (GiedriusS) [reply pkg/compact/compact.go:964]: This was good, old printf-debugging in e2e tests that I forgot to remove and hence not needed since I have figured out the problem since. The culprit was that I had forgotten to remove the `replica` label before trying to get the grouping key for the tests, and then the directory that I had been expecting was removed.

File: pkg/store/bucket_test.go
Comment (GiedriusS) [reply pkg/store/bucket_test.go:1057]: Yep :+1: it is starting to get a lot of arguments so we should refactor it into something more understandable

File: cmd/thanos/downsample.go
Comment (GiedriusS) [reply cmd/thanos/downsample.go:171]: Then it's not a problem because we have a consistent order of work with https://github.com/thanos-io/thanos/pull/3843 in place + https://github.com/thanos-io/thanos/blob/6877bd17aa3ebee44304aee4b2c6582be6f860cc/pkg/compact/compact.go#L312-L314 this exists already. This means that we are sure that after a retry, we are going to restart the work on the same blocks. If the blocks would disappear then accordingly we would get rid of them and then Thanos Compactor can continue its work. Does it make sense, @bwplotka ?

Comment (bwplotka) [reply cmd/thanos/downsample.go:171]: Makes sense 👍🏽

**Review by bwplotka — APPROVED** [review]
LGTM!

Rebase and feel free to merge 🤗

**Comment by GiedriusS** [comment]
:tada:

**Pull request merged**
